core/src/main/java/com/google/instrumentation/stats/Distribution.java

```diff
@@ -0,0 +1,236 @@
+ * histogram representing the distribution of those values across a specified set of histogram
+ * buckets.
+ *
+ * <p>The bucket boundaries for that histogram are described by {@link getBucketBoundaries}, which
+ * defines {@code getBucketBoundaries.size() + 1 (= N)} buckets. The boundaries for bucket index i
+ * are:
+ *
+ * <ul>
+ *   <li>[-infinity, bounds[i]) for i == 0
+ *   <li>[bounds[i-1], bounds[i]) for 0 &lt; i &lt; N-2
+ *   <li>[bounds[i-1], +infinity) for i == N-1
+ * </ul>
+ *
+ * <p>i.e. an underflow bucket (number 0), zero or more finite buckets (1 through N - 2, and an
+ * overflow bucket (N - 1), with inclusive lower bounds and exclusive upper bounds.
+ *
+ * <p>Note: If N = 1, there are no finite buckets and the single bucket is both the overflow and
+ * underflow bucket.
+ *
+ * <p>Although not forbidden, it is generally a bad idea to include non-finite values (infinities or
+ * NaNs) in the population of values, as this will render the {@code mean} meaningless.
+ */
+public final class Distribution {
+  private long count = 0; // The number of values in the population.
+  private double sum = 0.0; // The sum of the values in the population.
+  private Range range = new Range(); // Range of values in the population.
+  private final ArrayList<Double> bucketBoundaries; // Histogram boundaries; null means no histogram
+  private Long[] bucketCounts; // Counts for each histogram bucket
+
+  /**
+   * Constructs a new, empty {@link Distribution}.
+   *
+   * @return a new, empty {@code Distribution}.
+   */
+  public static final Distribution create() {
+    return new Distribution(null);
+  }
+
+  /**
+   * Constructs a new {@link Distribution} with specified {@code bucketBoundaries}.
+   *
+   * @param bucketBoundaries the boundaries for the buckets in the underlying {@code Distribution}.
+   * @return a new, empty {@code Distribution} with the specified boundaries.
+   * @throws NullPointerException if {@code bucketBoundaries} is null.
+   * @throws IllegalArgumentException if {@code bucketBoundaries} is not sorted, or has zero length.
+   */
+  public static final Distribution create(List<Double> bucketBoundaries) {
+    checkNotNull(bucketBoundaries, "bucketBoundaries");
+    checkArgument(bucketBoundaries.size() > 0, "Zero length bucket boundaries");
+    double lower = bucketBoundaries.get(0);
+    for (int i = 1; i < bucketBoundaries.size(); i++) {
+      double next = bucketBoundaries.get(i);
+      checkArgument(lower < next, "Bucket boundaries not sorted.");
+      lower = next;
+    }
+    return new Distribution(bucketBoundaries);
+  }
+
+  // Returns true if the distribution has histogram buckets.
+  private boolean hasBuckets() {
+    return (bucketBoundaries != null);
+  }
+
+  // Construct a new Distribution with given boundaries.
+  private Distribution(List<Double> bucketBoundaries) {
+    this.bucketBoundaries =
+        (bucketBoundaries == null) ? null : new ArrayList<Double>(bucketBoundaries.size());
```

Comment:
bucketBoundaries could be made immutable with `Collections.unmodifiableList(new ArrayList<Double>(bucketBoundaries))`.